

<!DOCTYPE html>


<html lang="en" data-content_root="" data-theme="light">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />
    <link href="_static/logo.svg" rel="icon" type="image/svg+xml">
    <title>evaluation package &#8212; GRouNdGAN 1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "light";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'evaluation';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="perturbation package" href="perturbation.html" />
    <link rel="prev" title="preprocessing package" href="preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="light">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="GRouNdGAN - Home"/>
    <script>document.write(`<img src="_static/logo.svg" class="logo__image only-dark" alt="GRouNdGAN - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://hub.docker.com/repository/docker/yazdanz/groundgan/" title="Docker" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-docker fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Docker</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://doi.org/10.1038/s41467-024-48516-6" title="Manuscript" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-solid fa-book fa-lg" aria-hidden="true"></i>
            <span class="sr-only">Manuscript</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN/actions" title="Docs Build" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><img src="https://github.com/Emad-COMBINE-lab/GRouNdGAN/actions/workflows/documentation.yaml/badge.svg?branch=master" class="icon-link-image" alt="Docs Build"/></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmarking.html">Benchmarking</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="modules.html">API References</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="factory.html">factory module</a></li>
<li class="toctree-l2"><a class="reference internal" href="gans.html">gans package</a></li>
<li class="toctree-l2"><a class="reference internal" href="layers.html">layers package</a></li>
<li class="toctree-l2"><a class="reference internal" href="main.html">main module</a></li>
<li class="toctree-l2"><a class="reference internal" href="networks.html">networks package</a></li>
<li class="toctree-l2"><a class="reference internal" href="custom_parser.html">custom_parser module</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html">preprocessing package</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">evaluation package</a></li>
<li class="toctree-l2"><a class="reference internal" href="perturbation.html">perturbation package</a></li>
<li class="toctree-l2"><a class="reference internal" href="sc_dataset.html">sc_dataset module</a></li>
<li class="toctree-l2"><a class="reference internal" href="tfrecord_loader.html">tfrecord_loader module</a></li>
</ul><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN/blob/master/docs/evaluation.rst?plain=1" target="_blank"
   class="btn btn-sm btn-source-file-button dropdown-item"
   title="Show source"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="btn__text-container">Show source</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN/edit/master/docs/evaluation.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/Emad-COMBINE-lab/GRouNdGAN/issues/new?title=Issue%20on%20page%20%2Fevaluation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/evaluation.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>evaluation package</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#submodules">Submodules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.MMD">evaluation.MMD module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD"><code class="docutils literal notranslate"><span class="pre">MMD</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.__init__"><code class="docutils literal notranslate"><span class="pre">MMD.__init__()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.squaredDistance"><code class="docutils literal notranslate"><span class="pre">MMD.squaredDistance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.gaussian_kernel"><code class="docutils literal notranslate"><span class="pre">MMD.gaussian_kernel()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.compute"><code class="docutils literal notranslate"><span class="pre">MMD.compute()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.data_quality">evaluation.data_quality module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.read_datasets"><code class="docutils literal notranslate"><span class="pre">read_datasets()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.plot_tSNE"><code class="docutils literal notranslate"><span class="pre">plot_tSNE()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.compute_distances"><code class="docutils literal notranslate"><span class="pre">compute_distances()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.compute_RF_AUROC"><code class="docutils literal notranslate"><span class="pre">compute_RF_AUROC()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.evaluate"><code class="docutils literal notranslate"><span class="pre">evaluate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.grn_inference">evaluation.grn_inference module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.get_imposed_grn"><code class="docutils literal notranslate"><span class="pre">get_imposed_grn()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.get_fake_grn"><code class="docutils literal notranslate"><span class="pre">get_fake_grn()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.compute_precision_at_k"><code class="docutils literal notranslate"><span class="pre">compute_precision_at_k()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.compute_PR"><code class="docutils literal notranslate"><span class="pre">compute_PR()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.evaluate"><code class="docutils literal notranslate"><span class="pre">evaluate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.lisi">evaluation.lisi module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.lisi.compute_lisi"><code class="docutils literal notranslate"><span class="pre">compute_lisi()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.lisi.compute_simpson"><code class="docutils literal notranslate"><span class="pre">compute_simpson()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation">Module contents</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="evaluation-package">
<h1>evaluation package<a class="headerlink" href="#evaluation-package" title="Permalink to this heading">#</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading">#</a></h2>
</section>
<section id="module-evaluation.MMD">
<span id="evaluation-mmd-module"></span><h2>evaluation.MMD module<a class="headerlink" href="#module-evaluation.MMD" title="Permalink to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="evaluation.MMD.MMD">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">evaluation.MMD.</span></span><span class="sig-name descname"><span class="pre">MMD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation/MMD.html#MMD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.MMD.MMD" title="Permalink to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Maximum Mean Discrepancy (MMD) class for computing distribution similarity
between real and generated samples using Gaussian kernels.</p>
<dl class="py method">
<dt class="sig sig-object py" id="evaluation.MMD.MMD.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/evaluation/MMD.html#MMD.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.MMD.MMD.__init__" title="Permalink to this definition">#</a></dt>
<dd><p>Initialize the MMD class with scale and weight parameters based on the median
nearest neighbor distance among real cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>real_cells</strong> (<em>np.ndarray</em>) – A NumPy array representing real cell data (cells x features).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="evaluation.MMD.MMD.squaredDistance">
<span class="sig-name descname"><span class="pre">squaredDistance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tensor</span></span></span><a class="reference internal" href="_modules/evaluation/MMD.html#MMD.squaredDistance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.MMD.MMD.squaredDistance" title="Permalink to this definition">#</a></dt>
<dd><p>Compute pairwise squared Euclidean distances between rows of X and Y.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – Input array of shape (n, d).</p></li>
<li><p><strong>Y</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – Input array of shape (m, d).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tensor of shape (n, m) representing squared distances.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tf.Tensor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="evaluation.MMD.MMD.gaussian_kernel">
<span class="sig-name descname"><span class="pre">gaussian_kernel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tensor</span></span></span><a class="reference internal" href="_modules/evaluation/MMD.html#MMD.gaussian_kernel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.MMD.MMD.gaussian_kernel" title="Permalink to this definition">#</a></dt>
<dd><p>Compute the multi-scale Gaussian kernel between two datasets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – Input array of shape (n, d).</p></li>
<li><p><strong>b</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – Input array of shape (m, d).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tensor of shape (n, m) representing the Gaussian kernel matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tf.Tensor</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="evaluation.MMD.MMD.compute">
<span class="sig-name descname"><span class="pre">compute</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tensor</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tensor</span></span></span><a class="reference internal" href="_modules/evaluation/MMD.html#MMD.compute"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.MMD.MMD.compute" title="Permalink to this definition">#</a></dt>
<dd><p>Compute the Maximum Mean Discrepancy (MMD) between two samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – First sample of shape (n, d).</p></li>
<li><p><strong>b</strong> (<em>np.ndarray</em><em> or </em><em>tf.Tensor</em>) – Second sample of shape (m, d).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The MMD score between the two distributions.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tf.Tensor</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-evaluation.data_quality">
<span id="evaluation-data-quality-module"></span><h2>evaluation.data_quality module<a class="headerlink" href="#module-evaluation.data_quality" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="evaluation.data_quality.read_datasets">
<span class="sig-prename descclassname"><span class="pre">evaluation.data_quality.</span></span><span class="sig-name descname"><span class="pre">read_datasets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cfg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ConfigParser</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/evaluation/data_quality.html#read_datasets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.data_quality.read_datasets" title="Permalink to this definition">#</a></dt>
<dd><p>Load real and simulated (fake) gene expression datasets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cfg</strong> (<em>ConfigParser</em>) – Parser for config file containing program params.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing:
- real_cells.X : NumPy array of real gene expression data
- fake_cells.X : NumPy array of simulated gene expression data, truncated to match real data row count</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[numpy.ndarray, numpy.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.data_quality.plot_tSNE">
<span class="sig-prename descclassname"><span class="pre">evaluation.data_quality.</span></span><span class="sig-name descname"><span class="pre">plot_tSNE</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fake</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/evaluation/data_quality.html#plot_tSNE"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.data_quality.plot_tSNE" title="Permalink to this definition">#</a></dt>
<dd><p>Perform t-SNE embedding on real and fake cell data and save a scatter plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real</strong> (<em>np.ndarray</em>) – A NumPy array of real cell data with shape (n_real_cells, n_features).</p></li>
<li><p><strong>fake</strong> (<em>np.ndarray</em>) – A NumPy array of fake/generated cell data with shape (n_fake_cells, n_features).</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Path to the directory where the t-SNE plot image will be saved.
If empty, the plot is not saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the 2D t-SNE embeddings of the real and fake data,
in the form (real_embedding, fake_embedding).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Tuple</em>[np.ndarray, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.data_quality.compute_distances">
<span class="sig-prename descclassname"><span class="pre">evaluation.data_quality.</span></span><span class="sig-name descname"><span class="pre">compute_distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fake_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/evaluation/data_quality.html#compute_distances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.data_quality.compute_distances" title="Permalink to this definition">#</a></dt>
<dd><p>Compute Euclidean and Cosine distances between the mean expression profiles of real and fake cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real_cells</strong> (<em>np.ndarray</em>) – A NumPy array representing real cell data (cells x features).</p></li>
<li><p><strong>fake_cells</strong> (<em>np.ndarray</em>) – A NumPy array representing fake cell data (cells × features).</p></li>
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – Axis along which to compute the mean expression (default is 0, meaning across cells), by default 0</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing:
- Euclidean distance between the mean expression profiles.
- Cosine distance between the mean expression profiles.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Tuple</em>[float, float]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.data_quality.compute_RF_AUROC">
<span class="sig-prename descclassname"><span class="pre">evaluation.data_quality.</span></span><span class="sig-name descname"><span class="pre">compute_RF_AUROC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fake_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="reference internal" href="_modules/evaluation/data_quality.html#compute_RF_AUROC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.data_quality.compute_RF_AUROC" title="Permalink to this definition">#</a></dt>
<dd><p>Compute the AUROC and plot the ROC curve of a Random Forest classifier distinguishing real from fake cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real_cells</strong> (<em>np.ndarray</em>) – A NumPy array representing real cell data (cells x features).</p></li>
<li><p><strong>fake_cells</strong> (<em>np.ndarray</em>) – A NumPy array representing fake/generated cell data (cells x features).</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Path to the directory where the ROC curve plot will be saved. The plot is saved as ‘RF.png’.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of principal components to retain during PCA, by default 50</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The area under the ROC curve (AUROC) for the classifier.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.data_quality.evaluate">
<span class="sig-prename descclassname"><span class="pre">evaluation.data_quality.</span></span><span class="sig-name descname"><span class="pre">evaluate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cfg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ConfigParser</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/evaluation/data_quality.html#evaluate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.data_quality.evaluate" title="Permalink to this definition">#</a></dt>
<dd><p>Assess the data quality of the simulated dataset.</p>
<ul class="simple">
<li><p>T-SNE plots of real vs simulated cells (jointly embedded and plotted).</p></li>
<li><dl class="simple">
<dt>Euclidean and Cosine distances between the centroids of real cells and simulated cells.</dt><dd><p>The centroid cell was obtained by calculating the mean along the gene axis (across all simulated or real cells).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Area under the receiver operating characteristic curve of a random forest in distinguishing real cells from simulated ones.</dt><dd><p>We perform dimensionality reduction using PCA to extract the top 50 PCs of each cell as the input features to the RF model.
The RF model is composed of 1000 trees and the Gini impurity was used to measure the quality of a split.</p>
</dd>
</dl>
</li>
<li><p>Maximum mean discrepancy (MMD) to estimate the proximity of high-dimensional distributions of real and simulated cells without
creating centroids.</p></li>
<li><dl class="simple">
<dt>Mean integration local inverse Simpson’s Index (miLISI).</dt><dd><p>iLISI captures the effective number of datatypes (real or simulated) to which datapoints of its local neighborhood belong.</p>
</dd>
</dl>
</li>
</ul>
<p>As a “control” (and to enable calibration of these scores), we also calculated these metrics using two halves of the reference test set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cfg</strong> (<em>ConfigParser</em>) – Parser for config file containing program params.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-evaluation.grn_inference">
<span id="evaluation-grn-inference-module"></span><h2>evaluation.grn_inference module<a class="headerlink" href="#module-evaluation.grn_inference" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="evaluation.grn_inference.get_imposed_grn">
<span class="sig-prename descclassname"><span class="pre">evaluation.grn_inference.</span></span><span class="sig-name descname"><span class="pre">get_imposed_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cfg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ConfigParser</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/evaluation/grn_inference.html#get_imposed_grn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.grn_inference.get_imposed_grn" title="Permalink to this definition">#</a></dt>
<dd><p>Load and process the imposed (ground-truth) GRN from a causal graph file.</p>
<p>This function reads the GRouNdGAN GRN and maps numeric gene indices to gene names
and constructs a structured representation of the GRN.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cfg</strong> (<em>ConfigParser</em>) – Configuration object providing the necessary file paths:
- “Data” -&gt; “train”: Path to real cells <cite>.h5ad</cite> file for gene name mapping.
- “Data” -&gt; “causal graph”: Path to the pickle file containing the ground truth GRN.
- “GRN Benchmarking” -&gt; “ground truth save path”: Optional output path for saving the ground truth GRN in csv format.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>imposed_grn</strong> (<em>pd.DataFrame</em>) – DataFrame with a single column “res” containing string-formatted edges (“TF -&gt; target”).</p></li>
<li><p><strong>imposed_TFs</strong> (<em>set</em>) – Set of transcription factors (TFs) involved in the imposed GRN.</p></li>
<li><p><strong>imposed_targets</strong> (<em>set</em>) – Set of target genes in the imposed GRN.</p></li>
<li><p><strong>imposed_edges</strong> (<em>list of str</em>) – List of edges as strings in the format “TF -&gt; target”.</p></li>
<li><p><strong>TFs</strong> (<em>list of str</em>) – List of unique transcription factors (from <cite>imposed_TFs</cite>), useful for downstream filtering.</p></li>
<li><p><strong>possible_edges_no</strong> (<em>int</em>) – Total number of possible TF-target combinations based on imposed GRN (<a href="#id1"><span class="problematic" id="id2">|TFs|</span></a> * <a href="#id3"><span class="problematic" id="id4">|targets|</span></a>).</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.grn_inference.get_fake_grn">
<span class="sig-prename descclassname"><span class="pre">evaluation.grn_inference.</span></span><span class="sig-name descname"><span class="pre">get_fake_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cfg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ConfigParser</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_TFs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_targets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">TFs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">possible_edges_no</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/evaluation/grn_inference.html#get_fake_grn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.grn_inference.get_fake_grn" title="Permalink to this definition">#</a></dt>
<dd><p>Construct a benchmark gene regulatory network (GRN) DataFrame using inferred and imposed edges.</p>
<p>This function loads an inferred GRN from a file, filters it based on imposed TFs
and targets, and augments it with all possible unreported edges (assigned zero importance).
This ensures that all theoretically possible edges are included for evaluation purposes (e.g., computing recall).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cfg</strong> (<em>ConfigParser</em>) – Configuration object providing the path to the inferred GRN file via the section
“GRN Benchmarking” -&gt; “grn to benchmark”.</p></li>
<li><p><strong>imposed_TFs</strong> (<em>Iterable</em><em>[</em><em>str</em><em>]</em>) – List of TFs that are considered “imposed” in the simulation and should be included as sources in the GRN.</p></li>
<li><p><strong>imposed_targets</strong> (<em>Iterable</em><em>[</em><em>str</em><em>]</em>) – List of target genes that are expected to be regulated in the benchmark.</p></li>
<li><p><strong>TFs</strong> (<em>Iterable</em><em>[</em><em>str</em><em>]</em>) – List of all possible TFs used to filter the GRN (must be valid source nodes).</p></li>
<li><p><strong>possible_edges_no</strong> (<em>int</em>) – Total number of possible edges in the benchmark GRN (used for validation/debugging).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame representing the benchmark GRN. Columns include:
- “TF”: Source transcription factor
- “target”: Target gene
- “importance”: Importance score (0 for non-inferred edges)
- “res”: String representation of the edge (“TF -&gt; target”)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.grn_inference.compute_precision_at_k">
<span class="sig-prename descclassname"><span class="pre">evaluation.grn_inference.</span></span><span class="sig-name descname"><span class="pre">compute_precision_at_k</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fake_grn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_grn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">TFs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numTFs_pergene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/evaluation/grn_inference.html#compute_precision_at_k"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.grn_inference.compute_precision_at_k" title="Permalink to this definition">#</a></dt>
<dd><p>Compute precision at k for inferred GRN against the imposed ground truth.
It plots the precision and baseline curves and optionally saves the figure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fake_grn</strong> (<em>pd.DataFrame</em>) – Inferred GRN with columns “TF”, “target”, and “res” (formatted as “TF -&gt; target”).</p></li>
<li><p><strong>imposed_grn</strong> (<em>pd.DataFrame</em>) – Ground truth GRN with a column “res” representing edges in “TF -&gt; target” format.</p></li>
<li><p><strong>TFs</strong> (<em>Union</em><em>[</em><em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>Set</em><em>[</em><em>str</em><em>]</em><em>]</em>) – List or set of valid transcription factors. Used to exclude predictions where a TF is treated as a target.</p></li>
<li><p><strong>numTFs_pergene</strong> (<em>int</em>) – Maximum number of top TFs to evaluate per gene.</p></li>
<li><p><strong>save_path</strong> (<em>str</em>) – Directory path where the precision-at-k plot should be saved. If empty, the plot is not saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.grn_inference.compute_PR">
<span class="sig-prename descclassname"><span class="pre">evaluation.grn_inference.</span></span><span class="sig-name descname"><span class="pre">compute_PR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fake_grn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_grn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_targets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">imposed_TFs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/evaluation/grn_inference.html#compute_PR"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.grn_inference.compute_PR" title="Permalink to this definition">#</a></dt>
<dd><p>Compute and plot the Precision-Recall (PR) curve for a predicted gene regulatory network (GRN)
against a ground-truth GRN.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fake_grn</strong> (<em>pd.DataFrame</em>) – Inferred GRN DataFrame with a column ‘res’ representing edges in the format “TF -&gt; target”.</p></li>
<li><p><strong>imposed_grn</strong> (<em>pd.DataFrame</em>) – Ground-truth GRN DataFrame with a column ‘res’ representing edges in the same format.</p></li>
<li><p><strong>imposed_edges</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of all ground-truth edges as strings (“TF -&gt; target”).</p></li>
<li><p><strong>imposed_targets</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em>) – Set of ground-truth target genes.</p></li>
<li><p><strong>imposed_TFs</strong> (<em>Set</em><em>[</em><em>str</em><em>]</em>) – Set of ground-truth transcription factors.</p></li>
<li><p><strong>save_path</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Directory path to save the precision-recall plot as “PR_curve.png”.
If None or empty, the plot is not saved.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.grn_inference.evaluate">
<span class="sig-prename descclassname"><span class="pre">evaluation.grn_inference.</span></span><span class="sig-name descname"><span class="pre">evaluate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cfg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ConfigParser</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/evaluation/grn_inference.html#evaluate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.grn_inference.evaluate" title="Permalink to this definition">#</a></dt>
<dd><p>Evaluate the reconstructed GRN against the ground truth GRN</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cfg</strong> (<em>ConfigParser</em>) – Parser for config file containing program params.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-evaluation.lisi">
<span id="evaluation-lisi-module"></span><h2>evaluation.lisi module<a class="headerlink" href="#module-evaluation.lisi" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="evaluation.lisi.compute_lisi">
<span class="sig-prename descclassname"><span class="pre">evaluation.lisi.</span></span><span class="sig-name descname"><span class="pre">compute_lisi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_colnames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="reference internal" href="_modules/evaluation/lisi.html#compute_lisi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.lisi.compute_lisi" title="Permalink to this definition">#</a></dt>
<dd><p>Compute the Local Inverse Simpson Index (LISI) for each label column in the metadata.</p>
<p>LISI measures the diversity of labels in the local neighborhood of each sample,
based on distances in the feature space <cite>X</cite>. It is commonly used to evaluate
dataset mixing (e.g., integration of multiple batches or experimental conditions).</p>
<dl class="simple">
<dt>A LISI score close to:</dt><dd><ul class="simple">
<li><p>1 indicates homogeneity (neighbors mostly from one label category),</p></li>
<li><p>N (number of categories) indicates high diversity (neighbors from all categories).</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em>) – Data matrix of shape (n_samples, n_features), representing the embedding space (e.g., PCA, UMAP).</p></li>
<li><p><strong>metadata</strong> (<em>pd.DataFrame</em>) – DataFrame containing categorical metadata for each sample (e.g., batch or cell type).</p></li>
<li><p><strong>label_colnames</strong> (<em>Iterable</em><em>[</em><em>str</em><em>]</em>) – List of column names in <cite>metadata</cite> for which to compute LISI.</p></li>
<li><p><strong>perplexity</strong> (<em>float</em><em>, </em><em>optional</em>) – Perplexity parameter influencing the neighborhood size, by default 30.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A matrix of shape (n_samples, len(label_colnames)) with LISI scores per label.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="evaluation.lisi.compute_simpson">
<span class="sig-prename descclassname"><span class="pre">evaluation.lisi.</span></span><span class="sig-name descname"><span class="pre">compute_simpson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distances</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indices</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Categorical</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_categories</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perplexity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tol</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1e-05</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="reference internal" href="_modules/evaluation/lisi.html#compute_simpson"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#evaluation.lisi.compute_simpson" title="Permalink to this definition">#</a></dt>
<dd><p>Compute the Simpson’s diversity index for each sample using a locally scaled probability distribution.</p>
<p>This function computes the effective number of categories in each sample’s neighborhood,
based on the similarity-weighted label distribution. It is internally used by <cite>compute_lisi</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>distances</strong> (<em>np.ndarray</em>) – Array of shape (n_neighbors, n_samples) with distances to each sample’s neighbors.</p></li>
<li><p><strong>indices</strong> (<em>np.ndarray</em>) – Array of shape (n_neighbors, n_samples) with indices of nearest neighbors for each sample.</p></li>
<li><p><strong>labels</strong> (<em>pd.Categorical</em>) – Categorical labels for all samples, aligned with the rows of <cite>X</cite>.</p></li>
<li><p><strong>n_categories</strong> (<em>int</em>) – Number of unique categories in the label.</p></li>
<li><p><strong>perplexity</strong> (<em>float</em>) – Perplexity value, used to set the target entropy for neighborhood probability distribution.</p></li>
<li><p><strong>tol</strong> (<em>float</em><em>, </em><em>optional</em>) – Tolerance for the entropy convergence, by default 1e-5.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of shape (n_samples,) with the Simpson’s index per sample.
Values closer to 1 indicate less diversity, higher values indicate more.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-evaluation">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-evaluation" title="Permalink to this heading">#</a></h2>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">preprocessing package</p>
      </div>
    </a>
    <a class="right-next"
       href="perturbation.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">perturbation package</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#submodules">Submodules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.MMD">evaluation.MMD module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD"><code class="docutils literal notranslate"><span class="pre">MMD</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.__init__"><code class="docutils literal notranslate"><span class="pre">MMD.__init__()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.squaredDistance"><code class="docutils literal notranslate"><span class="pre">MMD.squaredDistance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.gaussian_kernel"><code class="docutils literal notranslate"><span class="pre">MMD.gaussian_kernel()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.MMD.MMD.compute"><code class="docutils literal notranslate"><span class="pre">MMD.compute()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.data_quality">evaluation.data_quality module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.read_datasets"><code class="docutils literal notranslate"><span class="pre">read_datasets()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.plot_tSNE"><code class="docutils literal notranslate"><span class="pre">plot_tSNE()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.compute_distances"><code class="docutils literal notranslate"><span class="pre">compute_distances()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.compute_RF_AUROC"><code class="docutils literal notranslate"><span class="pre">compute_RF_AUROC()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.data_quality.evaluate"><code class="docutils literal notranslate"><span class="pre">evaluate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.grn_inference">evaluation.grn_inference module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.get_imposed_grn"><code class="docutils literal notranslate"><span class="pre">get_imposed_grn()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.get_fake_grn"><code class="docutils literal notranslate"><span class="pre">get_fake_grn()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.compute_precision_at_k"><code class="docutils literal notranslate"><span class="pre">compute_precision_at_k()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.compute_PR"><code class="docutils literal notranslate"><span class="pre">compute_PR()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.grn_inference.evaluate"><code class="docutils literal notranslate"><span class="pre">evaluate()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation.lisi">evaluation.lisi module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.lisi.compute_lisi"><code class="docutils literal notranslate"><span class="pre">compute_lisi()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evaluation.lisi.compute_simpson"><code class="docutils literal notranslate"><span class="pre">compute_simpson()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-evaluation">Module contents</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Yazdan Zinati
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023, Emad&#39;s COMBINE Lab: Yazdan Zinati, Abdulrahman Takiddeen, and Amin Emad.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>